sound/soc/sof/intel/hda-dai-ops.c

```diff
@@ -237,15 +237,32 @@ static unsigned int generic_calc_stream_format(struct snd_sof_dev *sdev,
 					       struct snd_pcm_substream *substream,
 					       struct snd_pcm_hw_params *params)
 {
+	struct snd_soc_pcm_runtime *rtd = asoc_substream_to_rtd(substream);
+	struct snd_soc_dai *codec_dai;
 	unsigned int format_val;
+	u32 ch_mask = 0;
+	int num_channels;
+	int codec_dai_id;
 
-	format_val = snd_hdac_calc_stream_format(params_rate(params), params_channels(params),
+	/*
+	 * if the multiple dais are handled by the same dailink, we may need to update the
+	 * stream channel count - the params are modified in soc-pcm based on the codec_ch_maps info
+	 */
+	for_each_rtd_codec_dais(rtd, codec_dai_id, codec_dai)
+		ch_mask |= rtd->dai_link->codec_ch_maps[codec_dai_id].ch_mask;
+
+	num_channels = hweight_long(ch_mask);
```

Comment:
this is more elegant, but counting inside the loop above might be faster and more compact

Reply:
no, we can't count inside the loop since some of the masks will overlap, so we might count the same channel multiple times. We have to check the mask after all the ORs are done.

Reply:
@ujfalusi each DAI has a channel mask, and we need to rebulld the dailink channel mask.

For amplifiers getting the same data, the dai masks are identical so the dailink mask will be the same as the dai masks.

For feedback capture, each dai has a different mask so the dailink will have a larger mask, the union of individual dai masks.